src/cargo/sources/git/utils.rs

```diff
@@ -557,29 +558,158 @@ fn with_authentication<T, F>(url: &str, cfg: &git2::Config, mut f: F)
     })
 }
 
-pub fn fetch(repo: &git2::Repository,
-             url: &str,
+pub fn fetch(repo: &mut git2::Repository,
+             url: &Url,
              refspec: &str,
              config: &Config) -> CargoResult<()> {
     if !config.network_allowed() {
         bail!("attempting to update a git repository, but --frozen \
                was specified")
     }
 
-    with_authentication(url, &repo.config()?, |f| {
+    // If we're fetching from github, attempt github's special fast path for
+    // testing if we've already got an up-to-date copy of the repository
+    if url.host_str() == Some("github.com") {
+        if let Ok(oid) = repo.refname_to_id("refs/remotes/origin/master") {
+            let mut handle = config.http()?.borrow_mut();
+            debug!("attempting github fast path for {}", url);
+            if github_up_to_date(&mut handle, url, &oid) {
+                return Ok(())
+            } else {
+                debug!("fast path failed, falling back to a git fetch");
+            }
+        }
+    }
+
+    // We reuse repositories quite a lot, so before we go through and update the
```

Comment:
Just to make sure we are on the same page, it's only the index repository which is frequently reused, because the crate repositories from git dependencies are write once?

Reply:
Hm, yeah, this part bother me:

>At the same time this commit also generalizes this logic, plus the existing fast
path github logic, to all git repositories and not just the index. That way all
git repositories can benefit from the "github fast path" as well as the
compaction steps.

I was under the impression that this is relevant only for index?

Reply:
I think the databases for git repos otherwise are updated quite a lot, but yes it's mainly the index and checkouts of git deps should be write-once.

Reply:
Oh but other repos on github will also benefit from the fast path "don't use git if we have an up-to-date copy" for cases where you don't have a `Cargo.lock` or you're running `cargo update`. The benefit is surely much more marginal, though.